pkg/reconciler/pipelinerun/resources/pipelineref.go

```diff
@@ -120,6 +120,20 @@ func (l *LocalPipelineRefResolver) GetPipeline(ctx context.Context, name string)
 	if err != nil {
 		return nil, nil, err
 	}
+	// Reapply defaults to the fetched Pipeline, since defaulting logic may have changed since it was created.
+	// For example, the Pipeline may have been created with an older server,
+	// and we may have added a new feature with a default value since then.
+	pipeline.SetDefaults(ctx)
+	// Skip validation of local Pipelines. Pipelines were validated when they were created.
+	// Local Pipelines have since been converted into the stored version, and we have no way of knowing
+	// what API version was used to create and validate them.
+	//
+	// For example, if enable-api-fields is "stable" and someone creates a v1beta1 Pipeline using beta features,
+	// this is perfectly valid. If the stored version of Pipeline is v1, the Pipeline is converted into v1,
+	// which doesn't allow beta features with enable-api-fields set to "stable". We should try to run the Pipeline anyway.
```

Comment:
This seems like a reason why we'd want to validate pipelines at reconcile, rather than skip it. 🤔

I don't think we can assume that validation was always ran, and even for the same API version the definition of valid may change between releases. In either case, if we're handed a spec that used to be valid but now isn't, the safer thing would be to validate it and respond accordingly rather than try to run it with potentially ambiguous outcomes.

Reply:
I agree this is a drawback of this PR, but I'm concerned that calling `pipeline.Validate` here is not going to actually fix https://github.com/tektoncd/pipeline/issues/6616. It would basically mean that swapping our storage version to v1 would break existing v1beta1 pipelines. Do you have suggestions for other approaches?